File: go/leetcode/5.最长回文子串.go
/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 *
 * https://leetcode-cn.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (24.38%)
 * Total Accepted:    38.2K
 * Total Submissions: 155.7K
 * Testcase Example:  '"babad"'
 *
 * 给定一个字符串 s，找到 s 中最长的回文子串。你可以假设 s 的最大长度为 1000。
 *
 * 示例 1：
 *
 * 输入: "babad"
 * 输出: "bab"
 * 注意: "aba" 也是一个有效答案。
 *
 *
 * 示例 2：
 *
 * 输入: "cbbd"
 * 输出: "bb"
 *
 *
 */
func findPalindrome(s string, left int, right int) string {
	j := left
	k := right
	l := len(s)
	ret := s[j : k+1]
	for {
		j--
		k++
		if j < 0 || k > l-1 {
			break
		}
		if s[j] != s[k] {
			break
		}
		ret = s[j : k+1]
	}
	return string(ret)
}

func longestPalindrome(s string) string {
	l := len(s)
	if l < 2 {
		return s
	}
	target := s[:1]
	for i := 0; i < l-1; i++ {
		if s[i] == s[i+1] {
			p := findPalindrome(s, i, i+1)
			if len(target) < len(p) {
				target = p
			}
		}
		if i-1 >= 0 && i+1 < l && s[i-1] == s[i+1] {
			p := findPalindrome(s, i-1, i+1)
			if len(target) < len(p) {
				target = p
			}
		}
	}
	return target
}

File: cpp/leetcode/5.最长回文子串.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] 最长回文子串
 *
 * https://leetcode-cn.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (34.41%)
 * Likes:    3723
 * Dislikes: 0
 * Total Accepted:    608.5K
 * Total Submissions: 1.8M
 * Testcase Example:  '"babad"'
 *
 * 给你一个字符串 s，找到 s 中最长的回文子串。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：s = "babad"
 * 输出："bab"
 * 解释："aba" 同样是符合题意的答案。
 *
 *
 * 示例 2：
 *
 *
 * 输入：s = "cbbd"
 * 输出："bb"
 *
 *
 * 示例 3：
 *
 *
 * 输入：s = "a"
 * 输出："a"
 *
 *
 * 示例 4：
 *
 *
 * 输入：s = "ac"
 * 输出："a"
 *
 *
 *
 *
 * 提示：
 *
 *
 * 1
 * s 仅由数字和英文字母（大写和/或小写）组成
 *
 *
 */

// @lc code=start
#include <iostream>
#include <string>
#include <utility>
class Solution {
public:
  std::string longestPalindrome(std::string s) {
    int start = 0, end = 0;
    for (size_t i = 0; i < s.size(); i++) {
      auto p1 = findPalindrome(s, i, i);
      auto p2 = findPalindrome(s, i, i + 1);
      if (p1.second - p1.first > end - start) {
        start = p1.first;
        end = p1.second;
      }
      if (p2.second - p2.first > end - start) {
        start = p2.first;
        end = p2.second;
      }
    }
    return s.substr(start, end- start + 1);
  }

  std::pair<int, int> findPalindrome(const std::string &s, int left,
                                     int right) {
    while (left >= 0 && right < s.size() && s[left] == s[right]) {
      left--;
      right++;
    }
    return {left + 1, right - 1};
  }
};
// @lc code=end

void print_ret(std::string ret) {
  std::cout << "ret: ";
  std::cout << ret << " ";
  std::cout << std::endl;
}

int main(int argc, const char **argv) {
  Solution s;
  auto ret = s.longestPalindrome("babad");
  print_ret(ret);
  return 0;
}